internal/repositories/devicedefinition/devicedefinition.go

```diff
@@ -0,0 +1,224 @@
+package devicedefinition
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	gmodel "github.com/DIMO-Network/identity-api/graph/model"
+	"github.com/DIMO-Network/identity-api/internal/helpers"
+	"github.com/DIMO-Network/identity-api/internal/repositories/base"
+	"github.com/vektah/gqlparser/v2/gqlerror"
+	"golang.org/x/exp/slices"
+	"net/http"
+	"net/url"
+	"path"
+	"strings"
+)
+
+// TokenPrefix is the prefix for a global token id for Device Definition.
+const TokenPrefix = "DD"
+
+type DeviceDefinitionRepository struct {
+	*base.Repository
+}
+
+func ToAPI(v *DeviceDefinitionTablelandModel) *gmodel.DeviceDefinition {
+	var result = gmodel.DeviceDefinition{}
+
+	result.ID = v.ID
+	result.Ksuid = &v.KSUID
+	result.Year = &v.Year
+	result.Model = &v.Model
+
+	if v.Metadata.DeviceAttributes != nil && len(v.Metadata.DeviceAttributes) > 0 {
+		for _, attr := range v.Metadata.DeviceAttributes {
+			result.Attributes = append(result.Attributes, &gmodel.DeviceDefinitionAttribute{
+				Name:  &attr.Name,
+				Value: &attr.Value,
+			})
+		}
+	}
+
+	return &result
+}
+
+func (r *DeviceDefinitionRepository) GetDeviceDefinition(ctx context.Context, by gmodel.DevicedefinitionBy) (*gmodel.DeviceDefinition, error) {
+	if by.Manufacturer == nil || *by.Manufacturer == "" {
+		return nil, gqlerror.Errorf("Provide exactly one `manufacturer`.")
+	}
+
+	if by.ID == nil || *by.ID == "" {
+		return nil, gqlerror.Errorf("Provide exactly one `ID`.")
+	}
+
+	tableName := "_80001_8439"
+	statement := fmt.Sprintf("SELECT * FROM %s WHERE id = '%s'", tableName, *by.ID)
+	queryParams := map[string]string{
+		"statement": statement,
+	}
+	var modelTableland []DeviceDefinitionTablelandModel
+
+	if err := r.QueryTableland(queryParams, &modelTableland); err != nil {
+		return nil, err
+	}
+
+	for _, item := range modelTableland {
+		return ToAPI(&item), nil
+	}
+
+	return nil, nil
+}
+
+func (r *DeviceDefinitionRepository) GetDeviceDefinitions(ctx context.Context, first *int, after *string, last *int, before *string, filterBy *gmodel.DevicedefinitionFilter) (*gmodel.DeviceDefinitionConnection, error) {
+	limit, err := helpers.ValidateFirstLast(first, last, base.MaxPageSize)
+	if err != nil {
+		return nil, err
+	}
+
+	if filterBy.Manufacturer == nil || *filterBy.Manufacturer == "" {
+		return nil, gqlerror.Errorf("Provide exactly one `manufacturer`.")
+	}
+
+	var conditions []string
+	if filterBy.Year != nil && (*filterBy.Year > 1980 && *filterBy.Year < 2999) {
+		conditions = append(conditions, fmt.Sprintf("year = %d", *filterBy.Year))
+	}
+	if filterBy.Model != nil && len(*filterBy.Model) > 0 {
```

Comment:
Use empty string check over 0 len for strings.
Also we should return an error if an empty string is given because that means the users explicitly set the value but set it to `""`
```suggestion
	if filterBy.Model != nil && *filterBy.Model != "" {
```